### PR TITLE
Record linearization in database

### LINKED ISSUE #2075: cylc db: record task linearisation
Various tools could be improved/extended/have missing features implemented if we had access to task linearisation (inheritance) information via the suite database.

Usages would include:
 * cylc suite-state working with FAMILIES
 * Rose Bush (or its cylc replacement) allowing grouping by and displaying of family information
 * Suite analytics tools to be able to work in terms of families rather than individual tasks

### PR DESCRIPTION
**Pull request by arjclark**
Closes #2075 

Adds a table recording task linearization to the suite database.

### CODE REVIEW

**Comment by hjoliver** [comment]
Much as I dislike creating more issues (we almost got down to 100 for a while...) when #2075 is closed by this PR, we should probably record the first bullet point of https://github.com/cylc/cylc/issues/2075#issue-193461447 in a new one?

File: tests/database/00-simple.t [hunk 1]
```diff
@@ -60,5 +60,10 @@ sqlite3 "${DB_FILE}" 'SELECT name, cycle, status FROM task_states ORDER BY name'
     >"${NAME}"
 cmp_ok "${TEST_SOURCE_DIR}/${TEST_NAME_BASE}/${NAME}" "${NAME}"
 
+NAME='select-linearization.out'
+sqlite3 "${DB_FILE}" 'SELECT namespace, linearization FROM linearization ORDER BY namespace' \
+    >"${NAME}"
+cmp_ok "${TEST_SOURCE_DIR}/${TEST_NAME_BASE}/${NAME}" "${NAME}"
+
```
Comment (hjoliver) [inline tests/database/00-simple.t]: Maybe rebase this branch to master and protect the `sqlite3` call as documented here: https://github.com/cylc/cylc/blob/master/bin/cylc-test-battery#L77 (else we might forget...)

**Comment by hjoliver** [comment]
Also wondering if `inheritance` might be a better name for the new table? (`linearization` is pretty obscure, for those not in the know).

**Comment by arjclark** [comment]
@hjoliver - rebased and changed to "inheritance" from "linearization"

**Comment by arjclark** [comment]
> we should probably record the first bullet point of #2075 (comment) in a new one

Yeah, that'd probably be for the best.

**Comment by matthewrmshin** [comment]
On reload, I wonder if it will be safer to delete everything and re-insert them all again.

**Comment by hjoliver** [comment]
@arjclark - 

> >    we should probably record the first bullet point of #2075 (comment) in a new one

> Yeah, that'd probably be for the best.

Just noted in #1984 that shared dependency objects might ultimately be the best route making family information available to external tools (I haven't thought this through in detail though...)

**Comment by hjoliver** [comment]
> I haven't thought this through in detail though

(huh, three different pronunciations of "ough" in one sentence :grimacing:)

**Comment by hjoliver** [comment]
As I recall we can't change inheritance structure on reload, so I don't think reinsert is needed (or has that changed?)

**Comment by matthewrmshin** [comment]
> As I recall we can't change inheritance structure on reload, so I don't think reinsert is needed (or has that changed?)

In which case, good to go.

**Pull request merged**